File: dfs/java/leetcode200_Number_of_Islands.java
//
//Given a 2d grid map of '1's (land) and '0's (water), count the number of islands. An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.
//
//Example 1:
//
//Input:
//11110
//11010
//11000
//00000
//
//Output: 1
//Example 2:
//
//Input:
//11000
//11000
//00100
//00011
//
//Output: 3

package wenxinjie2;

public class leetcode200_Number_of_Islands {
    public int numIslands(char[][] grid) {
        if (grid == null || grid.length == 0) return 0;
        int m = grid.length, n = grid[0].length;
        boolean[][] seen = new boolean[m][n];       
        int cnt = 0;        
        for (int i = 0; i < m ; i ++){
            for (int j = 0; j < n; j++){
                if ( !seen[i][j] && grid[i][j] == '1') {
                    cnt += 1;
                    helper(grid, seen,i,j);
                }
            }
        }
        return cnt;
    }
    
    public void helper(char[][] grid, boolean[][] seen, int i, int j){
            if (i < 0 || i > grid.length-1 || j < 0 || j > grid[0].length-1 || seen[i][j] || grid[i][j] == '0') return;
            seen[i][j] = true;
            helper(grid, seen, i+1, j);
            helper(grid, seen, i-1, j);
            helper(grid, seen, i, j+1);
            helper(grid, seen, i, j-1);
        }
}

//Time: O(m * n)
//Space: O(m * n)
//Difficulty: medium

File: dfs/python/leetcode200_Number_of_Islands.py

# Given a 2d grid map of '1's (land) and '0's (water), count the number of islands. An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

# Example 1:

# Input:
# 11110
# 11010
# 11000
# 00000

# Output: 1
# Example 2:

# Input:
# 11000
# 11000
# 00100
# 00011

# Output: 3

class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid: return 0
        m, n = len(grid), len(grid[0])
        seen = [[False for _ in range(n)] for _ in range(m)]
        
        def dfs(i,j):
            if  i < 0 or  i >= m or j < 0 or j >= n or seen[i][j] or grid[i][j] == '0':
                return 
            seen[i][j] = True
            for p in ((1,0), (-1,0), (0,-1), (0,1)):
                dfs(i+p[0], j+p[1])
        
        cnt = 0
        for i in range(m):
            for j in range(n):
                if not seen[i][j] and grid[i][j] == '1':
                    cnt += 1
                    dfs(i,j)
        return cnt

# Time: O(m * n)
# Space: O(m * n)
# Difficulty: medium
